src/daemon/daemon.cpp

```diff
@@ -2347,6 +2501,145 @@ catch (const std::exception& e)
     status_promise->set_value(grpc::Status(grpc::StatusCode::INTERNAL, e.what(), ""));
 }
 
+void mp::Daemon::snapshot(const mp::SnapshotRequest* request,
+                          grpc::ServerReaderWriterInterface<SnapshotReply, SnapshotRequest>* server,
+                          std::promise<grpc::Status>* status_promise)
+try
+{
+    mpl::ClientLogger<SnapshotReply, SnapshotRequest> logger{mpl::level_from(request->verbosity_level()),
+                                                             *config->logger,
+                                                             server};
+
+    const auto& instance_name = request->instance();
+    auto [instance_trail, status] = find_instance_and_react(operative_instances,
+                                                            deleted_instances,
+                                                            instance_name,
+                                                            require_operative_instances_reaction);
+
+    if (status.ok())
+    {
+        assert(instance_trail.index() == 0);
+        auto* vm_ptr = std::get<0>(instance_trail)->second.get();
+        assert(vm_ptr);
+
+        using St = VirtualMachine::State;
```

Comment:
I see this peppered in a few places, so I wonder if it makes sense adding it to `virtual_machine.h` and then dropping all of these?

Reply:
Hmm, I think it's best for the type in the include to remain spelled out and scoped to the VM. Calling a type `St` is only OK if it's used in the next line and contained in a small scope, IMO. And using declarations are seldom advisable in includes, since they drop namespace information for any code that includes them. 

The motivation is to help with formatting, to avoid having to repeat the long prefix with `off` and `stopped`, which requires breaking the line and makes it all harder to read. What we could do in the include would be to merge those two states, after ensuring that they are always treated the same. But I think that is for a separate PR.

Reply:
Ok, makes sense.  Let's leave it then.